### PR TITLE
Report status refinement and refactor

### PR DESCRIPTION
**Pull request by Pyifan**
The motivation of this change is to refine and refactor test report (mainly around the Status class) to make it more generic and easier to extend in future.

Major changes are:
- Separate RuntimeStatus(ready/running/finished) from Status(error/pass/fail/unknown/unstable). The former one is the runtime status of the Runnable object yet the latter run represents test result. Once they are separated, state transition and precedence calculation becomes easier and also makes better sense.
- Simplify status aggregation to only maintain precedence among a few primary status. And other status will be mapped to primary status, e.g xfail/xpass/skip --> unstable and fail/xpass-strict --> failed.
- Add a counter object to aggregate number of testcases of each status. Currently this is calculated in the UI but it makes better sense to have this statistic at backend.

All pytests passes, but I still need to make some UI changes which is not covered by test yet.
Also some clean-up might still be needed.

### CODE REVIEW

File: testplan/report/testing/base.py [hunk 1]
```diff
@@ -62,56 +63,80 @@
...
-        READY: READY,
-        RUNNING: RUNNING,
-        XFAIL: PASSED,
-        XPASS: PASSED,
-        None: None  # `status_override` can be None in merging report,
+        SKIPPED: UNSTABLE,
+        XFAIL: UNSTABLE,
+        XPASS: UNSTABLE,
+        UNSTABLE: UNSTABLE,
+        None: None  # `status_override` is default to None
```
Comment (ryan-collingham) [inline testplan/report/testing/base.py]: Again, we should not have a `None` status here. I think that the default status should be UNKNOWN (e.g. before tests are run, in interactive mode).

File: testplan/report/testing/schemas.py [hunk 2]
```diff
@@ -113,9 +114,10 @@ class TestGroupReportSchema(TestCaseReportSchema):
     category = fields.String()
     part = fields.List(fields.Integer, allow_none=True)
     fix_spec_path = fields.String(allow_none=True)
-    env_status = fields.String(allow_none=True)
+    # env_status = fields.String(allow_none=True)
```
Comment (ryan-collingham) [inline testplan/report/testing/schemas.py]: Why is this removed...? Please uncomment.

File: testplan/runnable/interactive/http.py [hunk 3]
```diff
@@ -319,15 +319,18 @@ def should_run(curr_status):
         setting the state of a running test to PASSED) is not allowed - only
         the server may make those transitions. A BadRequest exception will be
         raised if the requested status is not valid.
+
+        TODO: from api design perspective, should_run should take a curr_status
```
Comment (ryan-collingham) [inline testplan/runnable/interactive/http.py]: Could add a helper function `get_requested_status()` to handle extracting the new_status from the request JSON, then can call like:

```
if should_run(curr_status, get_requested_status()):
```

File: tests/functional/testplan/runnable/interactive/test_api.py [hunk 4]
```diff
@@ -470,6 +479,7 @@ def test_run_testcase(plan):
             raise_on_timeout=True,
         )
 
+# TODO: needs a bit more re-org around checking state transition
```
Comment (ryan-collingham) [inline tests/functional/testplan/runnable/interactive/test_api.py]: reorg how?

File: tests/functional/testplan/runnable/interactive/test_api.py [hunk 5]
```diff
@@ -424,7 +433,7 @@ def test_run_suite(plan):
 
     timing.wait(
         functools.partial(
-            _check_test_status, suite_url, "failed", updated_json["hash"]
+            _check_runtime_status, suite_url, "finished", updated_json["hash"]
```
Comment (ryan-collingham) [inline tests/functional/testplan/runnable/interactive/test_api.py]: I don't think we need `_check_test_status()` function anymore, `_check_runtime_status()` basically just replaces it.

Test logic should be:

1. Trigger test(s) to run
2. Poll for "runtime_status" to be FINISHED
3. Check that "status" is PASSED or FAILED as expected

File: tests/functional/testplan/testing/multitest/test_multitest_parts.py [hunk 6]
```diff
@@ -113,11 +113,14 @@ def test_multi_parts_invalid_parameter_1():
                   resource='MyPool')
 
     with log_propagation_disabled(TESTPLAN_LOGGER):
-        assert plan.run().run is False
+        result = plan.run()
```
Comment (ryan-collingham) [inline tests/functional/testplan/testing/multitest/test_multitest_parts.py]: ? Not sure that this change is necessary...

File: tests/functional/testplan/testing/multitest/test_xfail.py [hunk 7]
```diff
@@ -57,13 +57,18 @@ def test_xfail():
     )
     result = plan.run()
 
-    assert result.report.passed is False
+    assert result.report.failed is True
```
Comment (ryan-collingham) [inline tests/functional/testplan/testing/multitest/test_xfail.py]: can we just `assert result.report.failed` please?

File: testplan/report/testing/base.py [hunk 8]
```diff
@@ -62,56 +63,80 @@
...
+
+    READY = 'ready'
+    RUNNING = 'running'
+    FINISHED = 'finished'
+
+    STATUS_PRECEDENCE = (
+        READY,
+        RUNNING,
+        FINISHED,
+        None
```
[the remark this answers is not in the document]
Comment (Pyifan) [reply testplan/report/testing/base.py]: status_override might be of None value, so we need to merge None with None or None with other states.

File: testplan/report/testing/base.py [hunk 9]
```diff
@@ -62,56 +63,80 @@
...
+        UNKNOWN,    # black
+        PASSED,     # green
+        UNSTABLE,   # yellow
+        None        # `status_override` is default to None
     )
 
-    STATUS_PROPAGATION = {
+    # And we map detailed status to primary status when we propagate upward
+    # or decide an entry should be considered passed/failed/unstable/unknown
+    STATUS_MAPPING = {
```
[the remark this answers is not in the document]
Comment (Pyifan) [reply testplan/report/testing/base.py]: Yea, I think STATUS_CATEGORY might be better

File: testplan/report/testing/base.py [hunk 10]
```diff
@@ -124,14 +149,11 @@ def precedent(cls, stats, rule=STATUS_PRECEDENCE):
         :type rule: ``sequence``
         """
         return min(
-            [cls.STATUS_PROPAGATION[stat] for stat in stats],
+            [Status.STATUS_MAPPING[stat] for stat in stats],
```
[the remark this answers is not in the document]
Comment (Pyifan) [reply testplan/report/testing/base.py]: no, just changed code back and forth and made this unintentional change.

File: testplan/report/testing/base.py (hunk 1 above)
Comment (Pyifan) [reply testplan/report/testing/base.py]: status_override could be none, and when we merge 2 entries, we could be deciding the precedence between None and another status

File: testplan/runnable/interactive/http.py (hunk 3 above)
Comment (Pyifan) [reply testplan/runnable/interactive/http.py]: i'm not addressing this in this PR

File: tests/functional/testplan/runnable/interactive/test_api.py (hunk 4 above)
Comment (Pyifan) [reply tests/functional/testplan/runnable/interactive/test_api.py]: nothing, just a note to myself

File: tests/functional/testplan/runnable/interactive/test_api.py (hunk 5 above)
Comment (Pyifan) [reply tests/functional/testplan/runnable/interactive/test_api.py]: yea, i agree.

File: tests/functional/testplan/testing/multitest/test_multitest_parts.py (hunk 6 above)
Comment (Pyifan) [reply tests/functional/testplan/testing/multitest/test_multitest_parts.py]: Just for easy debugging

File: tests/functional/testplan/testing/multitest/test_xfail.py (hunk 7 above)
Comment (Pyifan) [reply tests/functional/testplan/testing/multitest/test_xfail.py]: sure

File: testplan/report/testing/base.py (hunk 1 above)
Comment (ryan-collingham) [reply testplan/report/testing/base.py]: If status_override is None, the status comes from the child entries no? I don't think there should be any case where the status of a report node is returned as `None`.

File: tests/functional/testplan/runnable/interactive/test_api.py (hunk 4 above)
Comment (ryan-collingham) [reply tests/functional/testplan/runnable/interactive/test_api.py]: if you are planning to reorg this test module can we discuss it first please

File: testplan/report/testing/base.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Pyifan) [reply testplan/report/testing/base.py]: Just being defensive.

File: testplan/report/testing/base.py (hunk 1 above)
Comment (Pyifan) [reply testplan/report/testing/base.py]: When we merge 2 entries (e.g multitest parts), we explicitly check status_override not the status property.

File: testplan/report/testing/schemas.py (hunk 2 above)
Comment (Pyifan) [reply testplan/report/testing/schemas.py]: done

File: tests/functional/testplan/runnable/interactive/test_api.py (hunk 4 above)
Comment (Pyifan) [reply tests/functional/testplan/runnable/interactive/test_api.py]: not in this PR
